File: diploma-server/src/test/java/org/arhor/diploma/core/BitSet32Test.java
package org.arhor.diploma.core;

import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

class BitSet32Test {

    @Test
    void shouldHaveAllValuesFalse() {
        // given
        var bitSet = new BitSet32();

        for (int i = 0; i < bitSet.size(); i++) {
            // when
            boolean value = bitSet.get(i);

            // then
            assertThat(value).isFalse();
        }
    }

    @Test
    void shouldHaveAllValuesTrue() {
        // given
        var bitSet = new BitSet32();

        // when
        for (int i = 0; i < bitSet.size(); i++) {
            bitSet.set(i, true);
        }

        // then
        for (int i = 0; i < bitSet.size(); i++) {
            assertThat(bitSet.get(i)).isTrue();
        }
    }

    @Test
    void shouldHaveOnlyOneValuesSet() {
        for (int indexToTest = 0; indexToTest < BitSet32.CAPACITY; indexToTest++) {
            // given
            var bitSet = new BitSet32();

            // when
            bitSet.set(indexToTest, true);

            // then
            for (int i = 0; i < bitSet.size(); i++) {
                if (i == indexToTest) {
                    assertThat(bitSet.get(i)).isTrue();
                } else {
                    assertThat(bitSet.get(i)).isFalse();
                }
            }
        }
    }

    @Test
    void shouldHaveSeparateIterators() {
        // given
        var bitSet = new BitSet32();

        // when
        var iterator1 = bitSet.iterator();
        var iterator2 = bitSet.iterator();

        // then
        assertThat(iterator1).isNotSameAs(iterator2);
    }

    @Test
    void shouldCorrectlyCreateBitSetFromNumber() {
        for (int indexToTest = 0; indexToTest < BitSet32.CAPACITY; indexToTest++) {
            // given
            var bitSet1 = new BitSet32();

            // when
            bitSet1.set(indexToTest, true);
            var bitSet2 = BitSet32.fromNumber(bitSet1.getValuesHolder());

            // then
            assertThat(bitSet2.getValuesHolder()).isEqualTo(bitSet1.getValuesHolder());

            for (int i = 0; i < BitSet32.CAPACITY; i++) {
                boolean value1 = bitSet1.get(i);
                boolean value2 = bitSet2.get(i);
                assertThat(value1).isEqualTo(value2);
            }
        }
    }
}

File: diploma-server/module-common/src/test/kotlin/org/arhor/diploma/commons/BitSet32Test.kt
package org.arhor.diploma.commons

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.MethodSource
import java.util.stream.IntStream

internal class BitSet32Test {

    @Test
    fun `should have all values set false by default`() {
        // given
        val bitSet = BitSet32()

        for (i in 0 until bitSet.size()) {
            // when
            val value = bitSet[i]

            // then
            assertThat(value).isFalse
        }
    }

    @Test
    fun `should correctly set all values true`() {
        // given
        val bitSet = BitSet32()

        // when
        for (i in 0 until bitSet.size()) {
            bitSet[i] = true
        }

        // then
        for (i in 0 until bitSet.size()) {
            assertThat(bitSet[i]).isTrue
        }
    }

    @Test
    fun `should create independent iterators`() {
        // given
        val bitSet = BitSet32()

        // when
        val iterator1 = bitSet.iterator()
        val iterator2 = bitSet.iterator()

        // then
        assertThat(iterator1).isNotSameAs(iterator2)
    }

    @ParameterizedTest(name = "{displayName} should modify only value with {0} index")
    @MethodSource("provideIndexesToTest")
    fun `BitSet32#set`(indexToTest: Int) {
        // given
        val bitSet = BitSet32()

        // when
        bitSet[indexToTest] = true

        // then
        for (i in 0 until bitSet.size()) {
            if (i == indexToTest) {
                assertThat(bitSet[i]).isTrue
            } else {
                assertThat(bitSet[i]).isFalse
            }
        }
    }

    @ParameterizedTest(name = "{displayName} should correctly create BitSet32 with only {0} index set")
    @MethodSource("provideIndexesToTest")
    fun `BitSet32#fromNumber`(indexToTest: Int) {
        // given
        val bitSet1 = BitSet32()

        // when
        bitSet1[indexToTest] = true
        val bitSet2 = BitSet32.fromNumber(bitSet1.valuesHolder)

        // then
        assertThat(bitSet2.valuesHolder).isEqualTo(bitSet1.valuesHolder)

        for (i in  0 until BitSet32.CAPACITY) {
            val value1 = bitSet1[i]
            val value2 = bitSet2[i]

            assertThat(value1).isEqualTo(value2)
        }
    }

    companion object {
        @JvmStatic
        @Suppress("UNUSED")
        private fun provideIndexesToTest(): IntStream {
            return IntStream.range(0, BitSet32.CAPACITY)
        }
    }
}
